src/cookies.js

```diff
@@ -83,29 +91,78 @@ function tryGetDocumentCookieNoInline(win) {
  */
 export function setCookie(win, name, value, expirationTime, opt_options) {
   checkOriginForSettingCookie(win, opt_options, name);
-  if (opt_options && opt_options.highestAvailableDomain) {
-    const parts = win.location.hostname.split('.');
-    let domain = parts[parts.length - 1];
-    for (let i = parts.length - 2; i >= 0; i--) {
-      domain = parts[i] + '.' + domain;
-      trySetCookie(win, name, value, expirationTime, domain);
-      if (getCookie(win, name) == value) {
-        return;
-      }
-    }
-  }
   let domain = undefined;
+  // Respect explicitily set domain over higestAvailabeDomain
   if (opt_options && opt_options.domain) {
     domain = opt_options.domain;
+  } else if (opt_options && opt_options.highestAvailableDomain) {
+    domain = getHighestAvailableDomain(win);
   }
   trySetCookie(win, name, value, expirationTime, domain);
 }
 
+/**
+ * Attemp to find the HighestAvailableDomain on
+ * @param {!Window} win
+ */
+export function getHighestAvailableDomain(win) {
+  // <meta name='amp-cookie-scope'>. Need to respect the meta first.
+
+  // Note: The same logic applies to shadow docs. Where all shadow docs are
+  // considered to be in the same origin. And only the <meta> from
+  // shell will be respected. (Header from shadow doc will be removed)
+  const metaTag = win.document.head.querySelector(
+    "meta[name='amp-cookie-scope']"
+  );
+
+  if (metaTag) {
+    // The content value could be an empty string. Return null instead
+    const cookieScope = metaTag.getAttribute('content') || '';
+    // Verify the validness of the amp-cookie-scope meta value
+    const sourceOrigin = getSourceOrigin(win.location.href);
+    // Verify the meta tag content value is valid
+    if (sourceOrigin.endsWith('.' + cookieScope)) {
+      return cookieScope;
+    } else {
+      // When the amp-cookie-scope value is invalid, fallback to the exact origin
+      // the document is contained in.
+      // sourceOrigin in the format of 'https://xxx or http://xxx'
+      return sourceOrigin.split('://')[1];
+    }
+  }
+
+  if (isProxyOrigin(win.location.href)) {
+    // return early for proxy origin.
+    return null;
+  }
+  const parts = win.location.hostname.split('.');
+  let domain = parts[parts.length - 1];
+  let testCookieName = TEST_COOKIE_NAME;
+  const counter = 0;
+  while (getCookie(win, testCookieName)) {
+    // test cookie name conflit, append counter to test cookie name
+    testCookieName = TEST_COOKIE_NAME + counter;
+  }
+  for (let i = parts.length - 2; i >= 0; i--) {
+    domain = parts[i] + '.' + domain;
+    // Try set a cookie for testing only, expire after 1 sec
+    trySetCookie(win, testCookieName, 'delete', Date.now() + 1000, domain);
+    if (getCookie(win, testCookieName) == 'delete') {
+      // Remove the cookie for testing
+      trySetCookie(win, testCookieName, 'delete', Date.now() - 1000, domain);
+      return domain;
+    }
+  }
+
+  // higestAvailableDomain cannot be found
+  return null;
```

Comment:
this method should not return null. it can always fallback to the exact source domain.

Reply:
What about `https://foo.bar.com`, where `foo.bar.com` is on the psl list?

Reply:
I suppose that's not going to happen, but I still want to return null for proxyOrigin w/o meta tag.

Reply:
in that case, we still need to write cookie on foo.bar.com right?

Reply:
honestly I don't know. I suppose browser won't let us.